internal/pkg/file/atomic.go

```diff
@@ -0,0 +1,341 @@
+/*
+Copyright 2024 k0s authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package file
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"runtime"
+)
+
+// The internal options for atomic file writes.
+type atomicOpts struct {
+	target      string
+	permissions fs.FileMode
+	uid, gid    int
+}
+
+func (o *atomicOpts) wantsChmod() bool {
+	return o.permissions.IsRegular()
+}
+
+type AtomicOpener struct{ atomicOpts }
+
+// Prepares to open a new [Atomic] for the file at the given target path.
+func AtomicWithTarget(target string) *AtomicOpener {
+	return &AtomicOpener{atomicOpts{
+		target:      target,
+		permissions: fs.ModeIrregular, // use this as an "unset" marker, see wantsChmod()
+		uid:         -1,
+		gid:         -1,
+	}}
+}
+
+// The desired permissions for the target.
+// Will rely on the umask if not called.
+func (o *AtomicOpener) WithPermissions(perm os.FileMode) *AtomicOpener {
+	o.permissions = perm.Perm()
+	return o
+}
+
+// The desired owner UID for the target file.
+// Will be owned by the current user if not called.
+// Will have no effect on Windows.
+func (o *AtomicOpener) WithOwner(uid int) *AtomicOpener {
+	o.uid = max(-1, uid)
+	return o
+}
+
+// The desired group ID for the target file.
+// Will be owned by the current user's group if not called.
+// Will have no effect on Windows.
+func (o *AtomicOpener) WithGroup(gid int) *AtomicOpener {
+	o.gid = max(-1, gid)
+	return o
+}
+
+// Open a new [Atomic] for writing. Writes to it will be unbuffered. It will be
+// backed by a hidden (i.e. its name will start with a dot), temporary file (it
+// will have a .tmp extension). If the returned Atomic gets closed without
+// calling [Atomic.Finish] before, the temporary file will be deleted without
+// touching the target. Use like so:
+//
+//	f, err := file.AtomicWithTarget("foo").Open()
+//	if err != nil {
+//		return err
+//	}
+//	defer f.Close()
+//	_, err = f.Write([]byte("I am atomic!"))
+//	if err != nil {
+//		return err
+//	}
+//	return f.Finish()
```

Comment:
There's Close() and Finish() here, but `Write` already calls `Do()` which does finish and close.

Reply:
The write you're referring to is is `Atomic.Write(...)` which implements `io.Writer`, not `AtomicOpener.Write()`, which is more like an atomic counterpart to `os.WriteFile().`